firedrake/adjoint/interpolate.py

```diff
@@ -0,0 +1,21 @@
+from firedrake.adjoint.blocks import InterpolateBlock
+from pyadjoint.tape import get_working_tape, stop_annotating, annotate_tape
+
+
+def annotate_interpolate(interpolate):
+    def wrapper(interpolator, *function, **kwargs):
+        annotate = annotate_tape(kwargs)
+        with stop_annotating():
+            output = interpolate(interpolator, *function, **kwargs)
+
+        if annotate:
+            sb_kwargs = InterpolateBlock.pop_kwargs(kwargs)
+            sb_kwargs.update(kwargs)
+            block = InterpolateBlock(interpolator, *function, **sb_kwargs)
```

Comment:
I believe the `InterpolateBlock` should be created before the actual `interpolate` call. This is a little fragile thing in pyadjoint, but it goes wrong when you have a dependency that is created by something that's not annotated. In that case the dependency, say a Function, does not have a checkpoint yet (as it's not the output of something), so we create a checkpoint when it's added as a dependency. Now where it goes wrong is that the dependency might also be the output - so you're interpolating an expression that depends on a Function into that same Function, which is not uncommon. In that case if you call interpolate first, the (input) value of the dependency that should be checkpointed is already clobbered with the output.

Reply:
Ahhhh! This might be the cause of the mesh bug!

Reply:
After making the changes, the mesh bug is still there.